cpp/src/sampling/negative_sampling_impl.cuh

```diff
@@ -412,16 +431,152 @@ std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> negativ
       // FIXME: We could oversample and discard the unnecessary samples
       // to reduce the number of iterations in the outer loop, but it seems like
       // exact_number_of_samples is an edge case not worth optimizing for at this time.
-      samples_in_this_batch = num_samples - current_sample_size;
+      samples_in_this_batch = total_samples - current_sample_size;
     } else {
       samples_in_this_batch = 0;
     }
   }
 
-  src.shrink_to_fit(handle.get_stream());
-  dst.shrink_to_fit(handle.get_stream());
+  srcs.shrink_to_fit(handle.get_stream());
+  dsts.shrink_to_fit(handle.get_stream());
+
+  if constexpr (multi_gpu) {
+    auto const& comm     = handle.get_comms();
+    auto const comm_size = comm.get_size();
+    auto const comm_rank = comm.get_rank();
+
+    // Randomly shuffle the samples so that each gpu gets their
+    // desired number of samples
+
+    if (!exact_number_of_samples) {
+      // If we didn't force generating the exact number of samples,
+      // we might have fewer samples than requested.  We need to
+      // accommodate this situation.  For now we'll just
+      // uniformly(-ish) reduce the requested size.
+      size_t total_extracted = host_scalar_allreduce(
+        handle.get_comms(), srcs.size(), raft::comms::op_t::SUM, handle.get_stream());
+      size_t reduction = total_samples - total_extracted;
+
+      while (reduction > 0) {
+        size_t est_reduction_per_gpu = (reduction + comm_size - 1) / comm_size;
+        for (size_t i = 0; i < samples_per_gpu.size(); ++i) {
+          if (samples_per_gpu[i] > est_reduction_per_gpu) {
+            samples_per_gpu[i] -= est_reduction_per_gpu;
+            reduction -= est_reduction_per_gpu;
+          } else {
+            reduction -= samples_per_gpu[i];
+            samples_per_gpu[i] = 0;
+          }
```

Comment:
I think this logic has a flaw.

Say reduction = 3 & comm_size  = 2.
Then, est_reduction_per_gpu = 2
If samples_per_gpu[] > 2, reduction = 3 - 2 - 2 = overflow!

Reply:
> reduce set_reduction_per_gpu to 1.

You mean est_reduction_per_gpu? This is set before the for loop. Once it is set to 2 in line 460, it will be 2 till the end of the for loop in line 416. Am I missing something?

Reply:
Oh, sorry, I missed the line 470.